### ISSUE TITLE
Import .kdbx file

### ISSUE THREAD

**Issue by tosim91** [keepassxreboot/keepassxc#1074]
In the FAQs it states you use the .kdbx database. Keepassxc does not let me import my kdbx database from KeePass2,v 2.32. Is there a workaround? Thanks for any assistance. I'm using Linux Mint 18.2MATE.
I installed Keepassxc from the " old Jessie deb" v, as the snap install did not run at all.

**Comment by phoerious** [comment]
You need to open it in KeePass2 and save it as KDBX3, not KDBX4, which is unsupported at the moment.

**Comment by affinityv** [comment]
From the RSS of releasess [1], KDBX4 didn't come in to play until 2.35 [2] and that is why Winblows machines that I update that are still using KP2 are being kept at 2.34.
I'm also hoping that KPXC doesn't force me to use KDBX4 for that reason.  I've had no troubles opening ANY database on KPXC to date without doing anything other than simply opening the files (with/without keyfiles if needed).

[1] https://sourceforge.net/p/keepass/news/feed?SetFreedomCookie
[2] https://keepass.info/news/n170109_2.35.html

**Comment by tosim91** [comment]
Thank you for the(amazingly) fast reply. I tried to do as you said, but it
keeps saving as a plain kdbx, and when I try (several times) for keepassxc
to open it, I get the following: Please see attachment. Once again, thanks
so much for the help.

**Comment by phoerious** [comment]
In case you are using a Yubikey, our implementation is incompatible with KeeChallenge. Everything else should work our of the box.

**Comment by tosim91** [comment]
Am not using Yubikey.

**Comment by phoerious** [comment]
Any particular error message?

**Comment by tosim91** [comment]
OK-I finally got it transferred. Had to make Save As: All files. Thank you guys for all your help. I'll close this now and "play" with kpxc, as I want to compare against kp2 which I've been using since it came out. Thanks again.

**Comment by mstarke** [comment]
@affinityv you can upgrade your Keepass installation to the current version since the update to KDBX4 is not done if you do not change anything that would require it. There are actually just a number of ways to make Keepass store your file in KDBX4 instead of KDBX3.1.
* You have a plugin that stores custom data for entries/groups
* You use anything besides AES for the key derivation
* You choose ChaCha as the encryption algorithm.
If you have no plugins installed that use the custom data and you do not touch the database settings regarding you encryption/key derivation Keepass will happily write a KDBX3.1 container.

**Comment by subdavis** [comment]
For anyone reading this in the future, I believe mstarke is mistaken above.

KeePass2 will gladly create a KDBXv3.1 container with ChaCha as the encryption algorithm as long as you use AES-KDF for key derivation.  

See issue referenced above (keeweb/kdbxweb#19) as well as the introduction section of https://keepass.info/help/kb/kdbx_4.html 

>Migration Phase. As not all major KeePass ports have finished adding support for KDBX 4 yet, for now KeePass 2.35 saves databases in this new format only when at least one of the following conditions is fulfilled:
> * AES-KDF is not selected as key derivation function (KDBX 3.1 only supports AES-KDF; any other key derivation function, like for instance Argon2, requires KDBX 4).
> * A plugin requests to store custom header data in the KDBX file.
> * A plugin requests to store custom data in an entry or a group.

**Comment by phoerious** [comment]
You can indeed save KDBX 3.1 files with ChaCha20. We also support exactly that and have tests to ensure this.

We save as KDBX 4 iff a KDF other than "AES-KDF (KDBX 3.1)" is used or custom plugin data exist. Unlike KeePass, we do support saving KDBX 4 with AES-KDF and for this allow selection between two different versions of it. KeePass can read these files just fine, but offers no way of explicitly saving in KDBX 4 format with AES-KDF. However, we do not recommend using AES-KDF at all.

**Comment by mstarke** [comment]
@subdavis you are completely right. Encryption is not a factor for file format decision since plugins can also add custom encryption algorithms for KDB3.1 container. Sorry for the misinformation. I mistakenly mixed up the random stream with the encryption. If KDBX4 is used, the inner random stream is ChaCha and not Salsa anymore.

**Comment by mstarke** [comment]
@phoerious KeePass does allow KDBX4 with AES-KDF as well. The UI has now way of doing this since you cannot add plugin data but if you use an XML file as import with custom data on entries or groups and then just leave AES-KDF and AES Encryption KeePass happily saves the file in KDBX4 format.

**Comment by phoerious** [comment]
Last time I checked there was only an option to use AES-KDF or Argon2, but no explicit switch for the file format.

**Comment by mstarke** [comment]
You comment suggest Keepass does not allow to use KDBX4 in combination with AES-KDF but it does but you are correct in that you cannot explicitly force KDBX4 without custom plugin data and AES-KDF.

**Comment by phoerious** [comment]
No, read again. KeePass can read those file (I tested that when I implemented it), but it didn't provide me with the option to convert a KDBX 3 file to KDBX 4 other than changing to Argon2.

**Comment by mstarke** [comment]
Correct. I didn't say anything to the contrary. KeePass decides implicitly what to use and if nothing requires KDBX4 it will fall back to KDB3.1. If anyone choses to save a "forced" KDBX4 file in KeePassXC and save it again on KeePass, it will get written as KDB3.1. That's not ideal and the only way to force KeePass to use KDB4 when using AES-KDF is via custom plugin data. If a file is prepared that way KeePass will then use KDBX4 as well but if you get rid of the plugin data, KDBX3.1 it is again. This I is what I meant by suggesting: KeePas can save KDBX4 with AES-KDF as well (if you add custom plugin data to force the KDBX4 format) but there is - as you correctly point out - no way of forcing KDB4.

I have no clue if this is something you might want to point out to users.

**Comment by phoerious** [comment]
Did you test if KeePass actually converts KDBX 4 files back to 3.1 if you don't explicitly open the database settings?
In any case, I would consider this low priority and a problem on KeePass's end. We won't recommend saving KDBX 3 anymore and switch to KDBX 4 by default in 2.4.

**Comment by mstarke** [comment]
If you take a look at the serialising code KeePass does not care in which format the db is at the moment it opens it, the decision on what to save is solely based on key derivation and custom data.

**Comment by mstarke** [comment]
I would be conservative with forcing KDBX 4 before KeePass does it since it might break some user's workflows since not all widely used clients support the new format. But I guess you had this discussion before you came to the conclusion to default to KDBX4 in 2.4.

From what I have experiences with the user-base of MacPass: There are a lot of novice users that have no clue that MacPass is just a port of another tool and do not know all the nitty-gritty details about the container nor the format an it's possibilities. The same might be true for KeepassXC. It's a balance to not scare them away and also give power users the tools they want to have.

**Comment by phoerious** [comment]
Existing databases will not be converted unless the user does so explicitly.

Switching to KDBX4 in 2.4 by default should be rather safe. With KeePassXC being one of the last ones to implement it, there aren't many KeePass ports left which don't support it. KeePassX still doesn't support it (and probably never will), but I don't see using KeePassX and KeePassXC in parallel as a valid use case.

We also have resonsilibities towards our users to provide them with modern cryptography standards. AES-KDF is simply not a good KDF and we cannot wait for everyone to catch up until we start pushing for Argon2. KDBX4 also does encrypt-than-MAC instead of MAC-then-encrypt, which is generally considered more secure and standard procedure, making KDBX4 not a perfect, but definitely the far superior format. So all in all, "KDBX4 by default" implies "secure by default", which is what we are aiming for.

Personally, I could even imagine phasing out support for KDBX3 and AES-KDF before KeePass does in a far-but-not-too-far-away future (we haven't discussed that yet, though). At the moment, KeePassXC is moving a lot faster than KeePass and while KeePass-compatibility is one of our top-priorities, secure, modern and fail-safe cryptography is even higher on the list.
